File: android/circlebox-sdk/src/main/kotlin/com/circlebox/sdk/CircleBox.kt
package com.circlebox.sdk

import android.content.Context
import java.io.File

/**
 * Public entrypoint for the CircleBox Android SDK.
 *
 * Start once during application launch, then add breadcrumbs at important user actions.
 */
object CircleBox {
    private val lock = Any()

    @Volatile
    private var appContext: Context? = null

    @Volatile
    private var runtime: CircleBoxRuntime? = null

    internal fun installContext(context: Context) {
        appContext = context.applicationContext
    }

    /**
     * Starts CircleBox with the given [config].
     *
     * Repeated calls after the first successful start are ignored.
     */
    @JvmStatic
    fun start(config: CircleBoxConfig = CircleBoxConfig()) {
        val rt = runtime()
        rt.start(config)
    }

    @JvmStatic
    fun start() {
        start(CircleBoxConfig())
    }

    /**
     * Adds a custom breadcrumb event to the in-memory ring buffer.
     */
    @JvmStatic
    fun breadcrumb(message: String, attrs: Map<String, String> = emptyMap()) {
        val rt = runtime()
        rt.breadcrumb(message, attrs)
    }

    @JvmStatic
    fun breadcrumb(message: String) {
        breadcrumb(message, emptyMap())
    }

    /**
     * Exports the latest report to files and returns absolute file handles.
     */
    @JvmStatic
    fun exportLogs(formats: Set<CircleBoxExportFormat> = setOf(CircleBoxExportFormat.JSON, CircleBoxExportFormat.CSV)): List<File> {
        val rt = runtime()
        return rt.exportLogs(formats)
    }

    @JvmStatic
    fun exportLogs(): List<File> {
        return exportLogs(setOf(CircleBoxExportFormat.JSON, CircleBoxExportFormat.CSV))
    }

    /**
     * Returns true when a crash report from a previous process run is available.
     */
    @JvmStatic
    fun hasPendingCrashReport(): Boolean {
        return runtime().hasPendingCrashReport()
    }

    /**
     * Clears the pending crash report file if it exists.
     */
    @JvmStatic
    fun clearPendingCrashReport() {
        runtime().clearPendingCrashReport()
    }

    private fun runtime(): CircleBoxRuntime {
        runtime?.let { return it }

        synchronized(lock) {
            runtime?.let { return it }

            val context = appContext
                ?: throw IllegalStateException("CircleBox context missing. Ensure CircleBoxInitProvider is merged or call installContext().")

            // Runtime is created lazily so apps can install context before the first API call.
            val created = CircleBoxRuntime(context)
            runtime = created
            return created
        }
    }
}

File: ios/CircleBoxSDK/Sources/CircleBoxSDK/CircleBox.swift
import Foundation

/// Public entrypoint for the CircleBox iOS SDK.
///
/// Call `start` once during app launch, then add breadcrumbs where useful.
public enum CircleBox {
    private static let runtime = CircleBoxRuntime()

    /// Starts CircleBox with the provided configuration.
    ///
    /// Repeated calls after the first one are ignored.
    public static func start(config: CircleBoxConfig = .default) {
        runtime.start(config: config)
    }

    /// Adds a custom breadcrumb to the ring buffer.
    ///
    /// - Parameters:
    ///   - message: Human-readable breadcrumb message.
    ///   - attrs: Optional structured context.
    public static func breadcrumb(_ message: String, attrs: [String: String] = [:]) {
        runtime.breadcrumb(message: message, attrs: attrs)
    }

    /// Records a screen view event into the ring buffer.
    ///
    /// Call this whenever the user navigates to a new screen so the crash
    /// timeline shows which view was active before the crash.
    ///
    /// - Parameters:
    ///   - name: Screen or view controller name (e.g. `"HomeScreen"`).
    ///   - attrs: Optional additional context.
    public static func screenView(_ name: String, attrs: [String: String] = [:]) {
        runtime.screenView(name: name, attrs: attrs)
    }

    /// Records a user interaction event into the ring buffer.
    ///
    /// Use this for high-signal actions such as button taps, form submissions,
    /// or gesture recognitions so the crash timeline captures what the user did
    /// immediately before the failure.
    ///
    /// - Parameters:
    ///   - actionType: Category of the action (e.g. `"tap"`, `"swipe"`, `"submit"`).
    ///   - target: Identifier of the element acted upon (e.g. `"LoginButton"`).
    ///   - attrs: Optional additional context.
    public static func userAction(_ actionType: String, target: String, attrs: [String: String] = [:]) {
        runtime.userAction(actionType: actionType, target: target, attrs: attrs)
    }

    /// Exports the latest report in the requested formats and returns file URLs.
    ///
    /// If a pending crash report exists from a previous launch, that report is exported.
    public static func exportLogs(formats: Set<CircleBoxExportFormat> = [.json, .csv]) throws -> [URL] {
        try runtime.exportLogs(formats: formats)
    }

    /// Returns `true` when a crash report from a previous process run exists.
    public static func hasPendingCrashReport() -> Bool {
        runtime.hasPendingCrashReport()
    }

    /// Deletes the pending crash report, if present.
    public static func clearPendingCrashReport() throws {
        try runtime.clearPendingCrashReport()
    }

    /// Returns a debug snapshot of the latest in-memory ring buffer events.
    ///
    /// This returns an empty list unless `enableDebugViewer` is enabled in config.
    public static func debugSnapshot(maxEvents: Int = 200) -> [CircleBoxEvent] {
        runtime.debugSnapshot(maxEvents: maxEvents)
    }

    /// Registers a non-blocking observer for realtime CircleBox events.
    ///
    /// Returns a token that can be passed to `removeEventObserver`.
    @discardableResult
    public static func addEventObserver(_ observer: @escaping (CircleBoxEvent) -> Void) -> UUID {
        runtime.addEventObserver(observer)
    }

    /// Removes a previously registered realtime event observer.
    public static func removeEventObserver(_ token: UUID) {
        runtime.removeEventObserver(token)
    }
}
